app/controllers/users_controller.rb

```diff
@@ -561,16 +545,30 @@ def is_local_username
         cannot_see[username] = cannot_see_reason if cannot_see_reason.present?
       end
 
+      usernames.each do |username|
+        group = groups[username]
+        next if group.blank?
+
+        if !guardian.can_see?(group)
+          cannot_see_reason = nil # do not leak private information
+        elsif topic.private_message? && !topic_allowed_group_ids.include?(group.id)
+          cannot_see_reason = :not_allowed
+        end
+
+        cannot_see_groups[username] = cannot_see_reason if cannot_see_reason.present?
+      end
+
       if usernames.include?(SiteSetting.here_mention) && guardian.can_mention_here?
         here_count = PostAlerter.new.expand_here_mention(topic.first_post, exclude_ids: [current_user.id]).size
       end
     end
 
     render json: {
       valid: users.keys,
-      valid_groups: groups,
-      mentionable_groups: mentionable_groups,
+      valid_groups: groups.keys,
```

Comment:
Was reading through the code and I wonder if this is a minor security flaw where we leak group names even if they are not public. I see how we take extra care to apply `guardian.can_see?(group)` for `cannot_see_groups` but not for `valid_groups`.

Reply:
Yeah, I have seen this one too. Is fixed in the refactor too.